File: aria_shell/modules/clock.py
from datetime import datetime

from gi.repository import Gtk

from aria_shell.module import AriaModule, GadgetRunContext
from aria_shell.config import AriaConfigModel
from aria_shell.gadget import AriaGadget
from aria_shell.gui import AriaPopover
from aria_shell.utils.logger import get_loggers
from aria_shell.utils import Timer


DBG, INF, WRN, ERR, CRI = get_loggers(__name__)


class ClockConfigModel(AriaConfigModel):
    format: str = '%H:%M'
    tooltip_format: str = '%A %d %B %Y'


class ClockModule(AriaModule):
    config_model_class = ClockConfigModel

    def __init__(self):
        super().__init__()
        self.timer: Timer | None = None

    def module_init(self):
        self.timer = Timer(1, self.timer_cb)

    def module_shutdown(self):
        if self.timer:
            self.timer.stop()
            self.timer = None

    def gadget_factory(self, ctx: GadgetRunContext) -> AriaGadget | None:
        conf: ClockConfigModel = ctx.config  # noqa
        instance = ClockGadget(conf)
        self.timer_cb(instance)  # perform a first update
        return instance

    def timer_cb(self, instance=None):
        now = datetime.now()
        if instance:
            instance.update(now)
        else:
            # print("Timer", now)
            for instance in self.gadgets:
                instance.update(now)
        return True


class ClockGadget(AriaGadget):
    def __init__(self, conf: ClockConfigModel):
        super().__init__('clock', clickable=True)
        self.conf = conf
        self.popover: AriaPopover | None = None

        self.label = Gtk.Label()
        self.append(self.label)

    def update(self, now: datetime):
        text = now.strftime(self.conf.format)
        self.label.set_text(text)

    def mouse_click(self, button: int):
        self.toggle_calendar()

    def toggle_calendar(self):
        if self.popover is None:
            calendar = Gtk.Calendar()
            self.popover = AriaPopover(self, calendar, self.on_popover_closed)
        else:
            self.popover.popdown()

    def on_popover_closed(self, _popover):
        self.popover = None

File: src/modules/clock.rs
use std::collections::HashMap;

use chrono::{DateTime, Local};

use crate::config::{ConfigSection, get_or};
use crate::module::{GadgetRunContext, Module};

/// Mirrors `aria_shell.modules.clock.ClockConfigModel`.
#[derive(Debug, Clone)]
pub struct ClockConfig {
    pub format: String,
    /// Parsed but UNUSED in this spike -- the calendar popover (Python's
    /// `AriaPopover`/`toggle_calendar`) is explicitly out of scope.
    #[allow(dead_code)]
    pub tooltip_format: String,
}

impl ConfigSection for ClockConfig {
    const SECTION: &'static str = "Clock";

    fn from_section(raw: &HashMap<String, String>) -> Self {
        Self {
            format: get_or(raw, "format", "%H:%M"),
            tooltip_format: get_or(raw, "tooltip_format", "%A %d %B %Y"),
        }
    }
}

/// Mirrors `ClockGadget`'s state (its `Gtk.Label` text), minus the
/// click-handling/popover bits (out of scope for this spike).
pub struct ClockState {
    format: String,
    text: String,
}

#[derive(Clone, Debug)]
pub enum Message {
    /// Mirrors `ClockModule.timer_cb(instance)`.
    Tick(DateTime<Local>),
}

pub struct ClockModule;

impl Module for ClockModule {
    type Config = ClockConfig;
    type Message = Message;
    type State = ClockState;

    fn gadget_factory(ctx: GadgetRunContext<ClockConfig>) -> ClockState {
        let now = Local::now();
        ClockState {
            text: now.format(&ctx.config.format).to_string(), // "perform a
            // first update", mirrors ClockModule.gadget_factory calling
            // self.timer_cb(instance) right after construction
            format: ctx.config.format,
        }
    }

    fn update(state: &mut ClockState, msg: Message) -> iced::Task<Message> {
        match msg {
            Message::Tick(now) => {
                state.text = now.format(&state.format).to_string();
            }
        }
        iced::Task::none()
    }

    fn view(state: &ClockState) -> iced::Element<'_, Message> {
        // Mirrors `Gtk.Label`; no click handler -- the calendar popover
        // is out of scope for this spike.
        iced::widget::text(state.text.clone()).into()
    }

    fn subscription(_state: &ClockState) -> iced::Subscription<Message> {
        // The real tick subscription lives at the top level (main.rs) and
        // broadcasts to every Clock instance -- see the doc comment on
        // `Module::subscription`.
        iced::Subscription::none()
    }
}
